### PR TITLE
Add dedicated linker section for firmware version string

### PR DESCRIPTION
**Pull request by bigdinotech**
Using a special macro DEFINE_FIRMWARE_VERSION_STRING(), you can
declare a string in C code which is guaranteed to be at 0x40033FC0

### CODE REVIEW

**Comment by bigdinotech** [comment]
@eriknyquist please review.
Replaces #6

File: include/arch/x86/linker.ld [hunk 1]
```diff
@@ -136,6 +136,17 @@ SECTIONS
 
 	__data_ram_end = .;
 
+#ifdef CONFIG_FIRMWARE_VERSION_STRING
+	SECTION_PROLOGUE(_VERSION_SECTION_NAME, (OPTIONAL),)
+	{
+	. = ALIGN(0x40033FC0);
```
Comment (eriknyquist) [inline include/arch/x86/linker.ld]: don't do this!!

You are setting the location counter to the next address on a 0x40033FC0 byte boundary. This is wrong for three reasons

1) This is a hack. There is already a way to declare a fixed load address for a section, just google "Fixed load address for linker section" to get the syntax

2) Due to the size of flash, ALIGN(0x40033FC0) *will* coincidentally set the location counter to exactly 0x40033FC0, but it is entirely dependent on the size of flash (i.e. you might get 0x0, or 0x80067F80 depending on the size of flash)

3) You're setting the location counter to 0x40033FC0. So all "romable" sections after this one will be after 0x40033FC0..... meaning we'll probably lose a bunch of stuff

**Review by eriknyquist — CHANGES_REQUESTED** [review]

File: include/arch/x86/linker.ld (hunk 1 above)
Comment (eriknyquist) [inline include/arch/x86/linker.ld]: Give the address a name, put it in linker-defs with all the other defs.
Surely you're aware by now that this is something we need to do?

Just always assume that you need to do this. We really don't need to spend time on trivial things like this in code review.

File: include/linker-defs.h [hunk 2]
```diff
@@ -32,6 +32,14 @@
 #include <toolchain.h>
 #include <sections.h>
 
+#ifdef CONFIG_FIRMWARE_VERSION_STRING
+#define VERSION_SECTION_SIZE 32
```
Comment (eriknyquist) [inline include/linker-defs.h:36]: Use this to define the start address. E.g. add a new define for the last address in Flash (or maybe it's already there). then the start address of the VERSION section is "FLASH_END - VERSION_SECTION_SIZE"

**Comment by bigdinotech** [comment]
@eriknyquist please review

File: include/arch/x86/linker.ld [hunk 3]
```diff
@@ -136,6 +136,15 @@ SECTIONS
 
 	__data_ram_end = .;
 
+#ifdef CONFIG_FIRMWARE_VERSION_STRING
+	_version_rom_start = FIRMWARE_FLASH_END - VERSION_SECTION_SIZE;
```
Comment (eriknyquist) [inline include/arch/x86/linker.ld]: put in linker-defs.h, with the others.....

File: include/arch/x86/linker.ld [hunk 4]
```diff
@@ -136,6 +136,15 @@ SECTIONS
...
+
+	_VERSION_SECTION_NAME _version_rom_start :
```
Comment (eriknyquist) [inline include/arch/x86/linker.ld]: Looks good. Let's use SECTION_PROLOGUE to match the rest of the file though

    SECTION_PROLOGUE(_VERSION_SECTION_NAME, _version_rom_start (OPTIONAL),)

**Comment by bigdinotech** [comment]
@eriknyquist please review

**Comment by bigdinotech** [comment]
@eriknyquist please review again haha

**Review by eriknyquist — APPROVED** [review]

**Comment by eriknyquist** [comment]
This won't compile for me (?). I made the same change to my own PR #6 (which has more info anyway), closing this one.

**Comment by eriknyquist** [comment]
replaced by #6

**Pull request closed without merging**